File: src/runner.ts
import { ChainReaction, ChainReactionOptions, XY } from "./chain_reaction";

export interface GameContext {
	width: ChainReaction["width"];
	height: ChainReaction["height"];
	grid: ChainReaction["grid"];

	canPlace: ChainReaction["canPlace"];
}

export interface Player {
	play(context: GameContext): XY | Promise<XY>;
}

interface RunnerOptions extends Omit<ChainReactionOptions, "players"> {
	players: Player[];
}

export class Runner {
	readonly game: ChainReaction;
	private players: Player[];
	private gameContext: GameContext;

	constructor(options: RunnerOptions) {
		const { width, height, players } = options;

		this.players = players;
		this.game = new ChainReaction({ width, height, players: players.length });
		this.gameContext = {
			width: width,
			height: height,
			grid: this.game.grid,
			canPlace: this.game.canPlace.bind(this.game),
		};
	}

	async run(): Promise<number> {
		while (this.game.isActive()) {
			const player = this.players[this.game.currentPlayer];

			const pos = await player.play(this.gameContext);
			await this.game.place(pos.x, pos.y);
		}

		const winner = this.game.winner();
		return winner;
	}
}

File: src/rust/src/runner.rs
use crate::array2d::Pos;
use crate::chain_reaction::ChainReaction;

pub trait Player {
  fn play(&mut self, game: &ChainReaction) -> Result<Pos, &'static str>;
}

pub struct Runner<F: Player> {
  width: usize,
  height: usize,
  game: ChainReaction,
  players: Vec<F>,
}

impl<F: Player> Runner<F> {
  pub fn new(width: usize, height: usize, players: Vec<F>) -> Result<Runner<F>, &'static str> {
    let game = ChainReaction::new(width, height, players.len())?;

    Ok(Runner {
      width,
      height,
      game,
      players,
    })
  }

  pub fn run(&mut self, times: u32) -> Result<Vec<usize>, &'static str> {
    let mut tally = vec![0; self.players.len()];
    for _ in 0..times {
      while self.game.active() {
        let player = self.game.current_player();
        let player_move = self.players[player].play(&self.game)?;
        self.game.place(&player_move)?;
      }

      let winner = self.game.winner()?;
      *tally.get_mut(winner).unwrap() += 1;
      self.game = ChainReaction::new(self.width, self.height, self.players.len())?;
    }
    Ok(tally)
  }
}
